File: src/bands/bands.controller.ts
import { Controller, Get, Post, Put, Delete, Body, Param, UseGuards } from '@nestjs/common';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { BandsService } from './bands.service';
import { CreateBandDto } from './dto/create-band.dto';
import { JwtAuthGuard } from '../auth/guards/jwt.guard';
import { SuperAdminGuard, AdminGuard } from '../auth/guards/admin.guard';
import { BandGuard } from '../common/guards/band.guard';
import { CurrentUser } from '../common/decorators/current-user.decorator';

@ApiTags('bands')
@ApiBearerAuth('JWT')
@Controller('bands')
export class BandsController {
  constructor(private readonly bands: BandsService) {}

  // -- Superadmin: manage all bands --

  @Get()
  @UseGuards(JwtAuthGuard, SuperAdminGuard)
  findAll() {
    return this.bands.findAll();
  }

  @Post()
  @UseGuards(JwtAuthGuard, SuperAdminGuard)
  create(@Body() dto: CreateBandDto) {
    return this.bands.create(dto);
  }

  @Delete(':id')
  @UseGuards(JwtAuthGuard, SuperAdminGuard)
  remove(@Param('id') id: string) {
    return this.bands.remove(id);
  }

  // -- Admin: manage own band --

  @Get('mine')
  @UseGuards(JwtAuthGuard, BandGuard)
  @ApiOperation({ summary: 'Obtener información de la banda actual' })
  getMyBand(@CurrentUser() user: any) {
    return this.bands.getMyBand(user.bandId);
  }

  @Put('mine')
  @UseGuards(JwtAuthGuard, BandGuard, AdminGuard)
  @ApiOperation({ summary: 'Actualizar nombre y logo de la banda' })
  updateMyBand(@CurrentUser() user: any, @Body() dto: { name?: string; logo?: string | null }) {
    return this.bands.updateMyBand(user.bandId, dto);
  }
}

File: spring-src/main/java/com/blackout/api/identity/infrastructure/web/BandsController.java
package com.blackout.api.identity.infrastructure.web;

import com.blackout.api.identity.application.service.BandsApplicationService;
import com.blackout.api.identity.infrastructure.web.dto.*;
import com.blackout.api.shared.domain.ForbiddenException;
import com.blackout.api.shared.infrastructure.security.BlackoutAuthentication;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/bands")
@Tag(name = "bands")
@SecurityRequirement(name = "bearerAuth")
public class BandsController {

    private final BandsApplicationService bandsService;

    public BandsController(BandsApplicationService bandsService) {
        this.bandsService = bandsService;
    }

    /** Superadmin: list all bands */
    @GetMapping
    @PreAuthorize("hasRole('SUPERADMIN')")
    public List<BandSummaryResponse> findAll() {
        return bandsService.findAll();
    }

    /** Superadmin: create band with initial admin user */
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    @PreAuthorize("hasRole('SUPERADMIN')")
    public CreateBandResponse create(@Valid @RequestBody CreateBandRequest req) {
        return bandsService.create(req);
    }

    /** Superadmin: delete band */
    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    @PreAuthorize("hasRole('SUPERADMIN')")
    public void remove(@PathVariable String id) {
        bandsService.remove(id);
    }

    /** Any authenticated user with a bandId: get own band */
    @GetMapping("/mine")
    public BandResponse getMyBand(BlackoutAuthentication auth) {
        requireBandId(auth);
        return bandsService.getMyBand(auth.getBandId());
    }

    /** Admin+: update own band */
    @PutMapping("/mine")
    @PreAuthorize("hasAnyRole('ADMIN', 'SUPERADMIN')")
    public BandResponse updateMyBand(BlackoutAuthentication auth,
                                      @RequestBody UpdateBandRequest req) {
        requireBandId(auth);
        return bandsService.updateMyBand(auth.getBandId(), req);
    }

    private void requireBandId(BlackoutAuthentication auth) {
        if (auth.getBandId() == null)
            throw new ForbiddenException("Token sin banda seleccionada. Usa /auth/select-band primero.");
    }
}
